src/field_theory/mv_polynomial.lean

```diff
@@ -0,0 +1,174 @@
+    sub_self, finset.prod_eq_zero_iff],
+  refine ⟨i, finset.mem_univ _, _⟩,
+  rw [finite_field.pow_card_sub_one_eq_one, sub_self],
+  rwa [(≠), sub_eq_zero],
+end
+
+lemma degrees_indicator (c : σ → α) :
+  degrees (indicator c) ≤ finset.univ.sum (λs:σ, add_monoid.smul (fintype.card α - 1) {s}) :=
+begin
+  rw [indicator],
+  refine le_trans (degrees_prod _ _) (finset.sum_le_sum $ assume s hs, _),
+  refine le_trans (degrees_sub _ _) _,
+  rw [degrees_one, ← bot_eq_zero, bot_sup_eq],
+  refine le_trans (degrees_pow _ _) (add_monoid.smul_le_smul_of_le_right _ _),
+  refine le_trans (degrees_sub _ _) _,
+  rw [degrees_C, ← bot_eq_zero, sup_bot_eq],
+  exact degrees_X _
+end
+
+lemma indicator_mem_restrict_degree (c : σ → α) :
+  indicator c ∈ restrict_degree σ α (fintype.card α - 1) :=
+begin
+  rw [mem_restrict_degree_iff_sup, indicator],
+  assume n,
+  refine le_trans (multiset.count_le_of_le _ $ degrees_indicator _) (le_of_eq _),
+  rw [← finset.sum_hom (multiset.count n)],
+  simp only [is_add_monoid_hom.map_smul (multiset.count n), multiset.singleton_eq_singleton,
+    add_monoid.smul_eq_mul, nat.cast_id],
+  transitivity,
+  refine finset.sum_eq_single n _ _,
+  { assume b hb ne, rw [multiset.count_cons_of_ne ne.symm, multiset.count_zero, mul_zero] },
+  { assume h, exact (h $ finset.mem_univ _).elim },
+  { rw [multiset.count_cons_self, multiset.count_zero, mul_one] }
+end
+
+section
+variables (α σ)
+def evalₗ : mv_polynomial σ α →ₗ[α] (σ → α) → α :=
+⟨ λp e, p.eval e,
+  assume p q, funext $ assume e, eval_add,
+  assume a p, funext $ assume e, by rw [smul_eq_C_mul, eval_mul, eval_C]; refl ⟩
+end
+
+section
+set_option class.instance_max_depth 50
+lemma evalₗ_apply (p : mv_polynomial σ α) (e : σ → α) : evalₗ α σ p e = p.eval e :=
+rfl
+end
+
+lemma map_restrict_dom_evalₗ : (restrict_degree σ α (fintype.card α - 1)).map (evalₗ α σ) = ⊤ :=
+begin
+  refine top_unique (submodule.le_def'.2 $ assume e _, mem_map.2 _),
+  refine ⟨finset.univ.sum (λn:σ → α, e n • indicator n), _, _⟩,
+  { exact sum_mem _ (assume c _, smul_mem _ _ (indicator_mem_restrict_degree _)) },
+  { ext n,
+    simp only [linear_map.map_sum, @pi.finset_sum_apply (σ → α) (λ_, α) _ _ _ _ _,
+      pi.smul_apply, linear_map.map_smul],
+    simp only [evalₗ_apply],
+    transitivity,
+    refine finset.sum_eq_single n _ _,
+    { assume b _ h,
+      rw [eval_indicator_apply_eq_zero _ _ h.symm, smul_zero] },
+    { assume h, exact (h $ finset.mem_univ n).elim },
+    { rw [eval_indicator_apply_eq_one, smul_eq_mul, mul_one] } }
+end
+end mv_polynomial
+
+namespace mv_polynomial
+universe u
+variables (σ : Type u) (α : Type u) [decidable_eq σ] [fintype σ] [discrete_field α] [fintype α]
+
+def R : Type u := restrict_degree σ α (fintype.card α - 1)
+
+instance R.acg : add_comm_group (R σ α) := by unfold R; apply_instance
+instance R.vs : vector_space α (R σ α) :=
+begin
+  show vector_space α (restrict_degree σ α (fintype.card α - 1)),
```

Comment:
Does `by delta R; apply_instance` not work?

Reply:
Thanks. `delta` and `dunfold` work.